### ISSUE TITLE
Subscribe to this issue to get requests for user feedback!

### ISSUE THREAD

**Issue by casey** [casey/just#1202]
Please subscribe to this issue if you'd like to receive notifications for requests for user feedback.

Situations often arise in which user feedback would be very helpful. Some examples are:

- Questions about whether a potential new feature is worth it or not.
- Questions about how a new feature should be implemented.
- Issues concerning existing features that I don't use.
- Issues concerning Windows, since I use Unix machines personally, and don't have a lot of Windows dev experience.
- Weird edge-cases where I can't figure out whether something is a bug and should be fixed, even if it might break existing justfiles.

If you're interested in helping out, subscribe to this issue! This issue is locked, so the only notifications that you'll get will be requests for help from me. I'll always link to another issue, and keep commentary out of this thread, so it should be focused.

**Comment by casey** [comment]
First request: Check out #1201 for a proposal for a backwards-compatible way of introducing otherwise breaking changes. Tell me what you think. Is this a good idea? Is there a better way to do it? Are there any pitfalls I haven't thought of? Will it be super annoying if people have to put `set edition := 2` to get better defaults?

**Comment by casey** [comment]
Second request: Initially, `just` used `sh` on all platforms. This required installing `sh` on Windows, but made justfiles work the same everywhere. Later, the `set windows-powershell` setting was introduced, which allowed using `powershell.exe` on Windows but `sh` everywhere else. PowerShell and `sh` are often compatible, so this was a good way of writing cross-platform `justfiles`. Later, Microsoft introduced a new version of PowerShell as `pwsh.exe`. The `set windows-powershell` setting can't be changed to use `pwsh.exe`, because the two version are incompatible, and some `justfile`s using `set windows-powershell` might break. So, it is now proposed that we adde a new setting `set windows-shell := […]` that allows setting an arbitrary shell for Windows. This allows users to opt-in to whatever shell they like on Windows, and is probably what we should have done from the beginning. I'm pretty sure this is a good change, but I don't use Windows, so I'm always nervous about things like this. Comment and let me know what you think! See the PR here https://github.com/casey/just/pull/1198 and the issue with comments here https://github.com/casey/just/issues/1190.

**Comment by casey** [comment]
Third request: [A proposal for a new recipe type](https://github.com/casey/just/issues/1413), which solves a number of issues and inconsistencies, and could be made the default recipe type in a future edition.

**Comment by casey** [comment]
Fourth request! Just now has modules and imports.

Imports load a source file and include all items in the source file in the current justfile.

If a `justfile` contains:

```just
import 'foo.just'
```

And `foo.just` contains:

```just
foo:
  @echo FOO
```

Then running  `just foo` will print `FOO`.

Modules create independent submodules, whose recipes can be invoked as subcommands:

If a `justfile` contains:

```just
mod foo
```

And `foo.just` contains:

```just
bar:
  @echo BAR
```

Then running  `just foo bar` will print `BAR`.

Imports are stable, but modules are currently unstable, so you'll need the `--unstable` flag to use them.

Modules are a big feature, so I'm hoping to get plenty of testing and feedback before stabilizing them. Try them out and let me know what you think! You can see planned features and leave feedback in #929.

**Comment by casey** [comment]
I want to give a final call for testing and feedback for the modules feature. All bugs I know of have been fixed, and I added a `source_file()` and `source_directory()` for getting the file of the current source file, and the directory that contains it, respectively.

I think it's pretty much ready to go, and I'll try to stabilize them in a month, assuming nothing goes wrong. Check out the [tracking issue](https://github.com/casey/just/issues/929) to leave your comments!

Also, while I have your attention, we just [released 1.27.0](https://github.com/casey/just/releases/tag/1.27.0), which has a ton of new features and functions, so check that out too!

**Comment by casey** [comment]
I've landed a few PRs that improve `--fmt` output recently:

- #3275
- #3274
- #3268
- #3282

I now have a bit of a dilemma. I'm starting to think that formatting output is good enough that the feature could be stabilized. However, I think that formatting is a feature which is likely to be subject to tweaking over the long term. Better ways to format the same thing (like breaking long expressions onto multiple lines), and fixing random bugs.

Given all that, there are a few options:

1. `--fmt` is permanently unstable
2. `--fmt` is stable but can change in a minor just release and isn't bound by just's normal stability guarantee (y in `x.y.z`)
3. `--fmt` is stable but we can't ever improve it

I think that 2. is probably the best option, but it's a departure from the usual extremely strong stability guarantees. If we did 2., I would probably still try to avoid frequent changes, which, most annoyingly would break CI jobs that use `--fmt --check`, maybe keeping them to once or twice per year.

Let me know what you think in #868!

**Comment by casey** [comment]
Ten years ago I started the Rust version of `just` with the auspicious commit message "Initial commit. Broken." It's crazy to me that it's been so long, and that I still enjoy working on it. Here's to ten more years!

I just (lmao, never gets old) released [1.53.0](https://github.com/casey/just/releases/tag/1.53.0), which has an absolutely massive new feature: lists.

It is gated behind the `set lists` setting and currently unstable, so you'll also need `set unstable` or `export JUST_UNSTABLE=1`.

Previously, all values were strings. With `set lists`, all values are lists of strings. Single strings like `"foo"` are actually `["foo"]`.

This may seem somewhat odd, but I hope it's a good choice that fits `just` well. Over the years, there have been requests for many features which were hard to implement with strings. Things like correctly quoting lists of arguments, dispatching arguments to dependencies, and, of course, manipulating lists of paths and flags, which are very common in `just`.

This can all be implemented easily if all values are lists of strings. This preserves the simplicity of `just`'s type system, since all values still have the same type.

It's certainly somewhat odd that strings are lists of length one, but it's been used in at least two other places: the [rc shell](https://en.wikipedia.org/wiki/Rc_(Unix_shell)) and the [fish shell](https://en.wikipedia.org/wiki/Fish_(Unix_shell)), specifically to solve the kinds of same kinds of quoting and splitting problems that `just` has.

Lists do not yet work in all contexts. In particular, lists do not work with many built-in functions, since it's not obvious how many of them should operate with lists. Feedback on how built-in functions should work with lists, and on the feature in general, is most welcome! Please feel free to open an issue, or leave a comment in the [tracking issue](https://github.com/casey/just/issues/3377).

Two headline features are properly quoting variadic arguments:

```just
set unstable
set lists

foo *args:
  echo {{ quote(args) }}
```

And mapping variadic arguments to dependencies, especially in parallel:

```just
set unstable
set lists

# call `compile` with each element of `args` in parallel
[parallel]
build *args: *(compile *args)

compile target:
```

But there is a a ton of functionality, all documented in the [lists section of the readme](https://github.com/casey/just#lists), to avail yourself of.

Try it out and let me know what you think!
